src/inc/til/rle.h

```diff
@@ -0,0 +1,1066 @@
+
+                if (totalLength > position)
+                {
+                    // If we've just passed up the requested position with the length we added, break early
+                    break;
+                }
+
+                runPos++;
+            } while (runPos < _list.end());
+
+            // we should have broken before falling out the while case.
+            // if we didn't break, then this ATTR_ROW wasn't filled with enough attributes for the entire row of characters
+            FAIL_FAST_IF(runPos >= _list.end());
+
+            // The remaining iterator position is the position of the attribute that is applicable at the position requested (position)
+            // Calculate its remaining applicability if requested
+
+            // The length on which the found attribute applies is the total length seen so far minus the position we were searching for.
+            FAIL_FAST_IF(!(totalLength > position)); // The length of all attributes we counted up so far should be longer than the position requested or we'll underflow.
+
+            applies = totalLength - position;
+            FAIL_FAST_IF(!(applies > 0)); // An attribute applies for >0 characters
+            // MSFT: 17130145 - will restore this and add a better assert to catch the real issue.
+            //FAIL_FAST_IF(!(attrApplies <= _size)); // An attribute applies for a maximum of the total length available to us
+
+            return runPos;
+        }
+
+        template<class Iter>
+        void _merge(Iter first,
+                    Iter last,
+                    const S startIndex)
+        {
+            // Definitions:
+            // Existing Run = The run length encoded color array we're already storing in memory before this was called.
+            // Insert Run = The run length encoded color array that someone is asking us to inject into our stored memory run.
+            // New Run = The run length encoded color array that we have to allocate and rebuild to store internally
+            //           which will replace Existing Run at the end of this function.
+            // Example:
+            // _size = 10.
+            // Existing Run: R3 -> G5 -> B2
+            // Insert Run: Y1 -> N1 at startIndex = 5 and endIndex = 6
+            //            (rgInsertAttrs is a 2 length array with Y1->N1 in it and cInsertAttrs = 2)
+            // Final Run: R3 -> G2 -> Y1 -> N1 -> G1 -> B2
+
+            const auto newItemsTotalCoverage = std::accumulate(first, last, (S)0, [](S value, auto& item) -> S {
+                return value + gsl::narrow_cast<S>(item.second);
+            });
+
+            const auto newItemsSize = std::distance(first, last);
+
+            // If the insertion size is 1, do some pre-processing to
+            // see if we can get this done quickly.
+            if (newItemsSize == 1)
+            {
+                // Get the new color attribute we're trying to apply
+                const T NewAttr = first->first;

```

Comment:
Boy do I wish there's `NamedTuple` in C++

Reply:
Yeah uh I got it backwards a few times in this PR while testing. But I'm doing `T, S` consistently now. So all `->first`  should be the `T` type and all `->second` should be the `S` type.

I did temporarily have a custom structure to hold the pair so it would be named and have a custom debug print string... but then when I went to go start implementing all of its operators correctly...I realized I should just shove it in the `std::pair<_Ty1, _Ty2>` template and let it do all the boilerplate for me.